### PR TITLE
Allow for filtering by frontmatter properties

### LINKED ISSUE #257: [Feature request] - Map View query mechanism should be able to filter by properties
As I mentioned in #255, I think it would be very useful to allow queries such as `[propertyname:propertyvalue]` in the Map View Filters section. It should be in a way that resembles the Obsidian query syntax, so, e.g., if property `foo` has value `[[Bar]]` or a list of values, one of which is `[[Bar]]`, then `[foo:Bar]` should consider it.

If this is too much to implement, perhaps it would be easier to allow `linkedto:` to also read from property values (e.g., if the property `foo` has value `[[Bar]]` or a list of values, one of which is `[[Bar]]`, then `linkedto:"Bar"` should consider it.

### PR DESCRIPTION
**Pull request by zakj**
Support for `[property-name:property-value]` to query frontmatter. A quoted name or quoted value indicates an exact match on that name or value, and unquoted name/value does substring matching (this is how Obsidian handles it). Frontmatter property values are coerced to strings, so it's possible to match boolean and numerical properties. Date properties are already represented as strings in the file cache.

Fixes #257.

### CODE REVIEW

File: src/query.ts [hunk 1]
```diff
@@ -36,12 +36,15 @@ export class Query {
...
+        // 4. Replace ["property":"value"] with single quotes to avoid parser complaints.
         let newString = queryString
             .replace(regex.TAG_NAME_WITH_HEADER_AND_WILDCARD, '"tag:$1"')
             .replace(regex.PATH_QUERY_WITH_HEADER, '"path:$1"')
             .replace(regex.LINKEDTO_QUERY_WITH_HEADER, '"linkedto:$1"')
             .replace(regex.LINKEDFROM_QUERY_WITH_HEADER, '"linkedfrom:$1"')
-            .replace(regex.NAME_QUERY_WITH_HEADER, '"name:$1"');
+            .replace(regex.NAME_QUERY_WITH_HEADER, '"name:$1"')
+            .replace(/^\[(")(.+?)\1:/, "['$2':")
+            .replace(/:(")(.+)?\1\]/, ":'$2']");
```
Comment (zakj) [inline src/query.ts:47]: This is admittedly terrible; we may be stretching the limits of `boon-js` at this point, but I don't see a better way to handle it.

File: src/query.ts [hunk 2]
```diff
@@ -154,10 +157,58 @@ export class Query {
...
+// Return whether the given string was quoted, and the unquoted value.
+function unquote(s: string): [boolean, string] {
+    // Match any string, but only capture the first group on balanced quotes.
+    const match = s.match(/^(['"])?(.*)\1$/);
+    if (!match) {
+        throw new Error('Unexpected regex failure when unquoting: ' + s);
+    }
+    return [Boolean(match[1]), match[2]];
+}
+
```
Comment (zakj) [inline src/query.ts:211]: Let me know if these belong in `utils.ts` or elsewhere, but I figured I'd put them close to where they're used.

Comment (esm7) [reply src/query.ts:212]: Agreed, it's fine here :)

**Review by esm7 — APPROVED** [review]
Well-written and clean as much as the infrastructure can allow, I really have nothing to complain about :)

**Comment by lamelas** [comment]
Hello! This seems to be an issue with boon-js but when I use a property value with a space, I get an "Invalid Token" exception from the parse method and `this.queryRpn` ends up as null.

https://github.com/zakj/obsidian-map-view/blob/acc8ced7612f582d637a906662b831b3c1beec40/src/query.ts#L29-L31

**Pull request merged**